### DISCUSSION TITLE
Cosmos DB Identity container indexes

### DISCUSSION THREAD

**Discussion by etriebe** [MoonriseSoftwareCalifornia/AspNetCore.Identity.CosmosDb#27] [General]
Azure is recommending I create the following indexes: 

- "/Discriminator   ASC","/UserName ASC"

I went into Cosmos DB data explorer, selected the Identity container and tried to add the following indexing policy. 

```json
{
    "indexingMode": "consistent",
    "automatic": true,
    "includedPaths": [
        {
            "path": "/Discriminator"
        },
        {
            "path": "/UserName"
        }
    ],
    "excludedPaths": [
        {
            "path": "/*"
        }
    ]
}
```

And got this error.

```
Failed to update container Identity:
 Message: {"code":"BadRequest","message":"Message: {\"Errors\":[\"The indexing path '\\/Discriminator' could not be accepted, failed near position '14'. Please ensure that the path is a valid path. Common errors include invalid characters or absence of quotes around labels.\"]}\r\nActivityId: f6cb1f05-e977-4b15-b38b-6271b66e21ef, Request URI: /apps/f50faba1-c002-4cef-b681-f213f30661a3/services/1e1ba7eb-405b-4be7-a472-3a08a060b169/partitions/350ba870-568a-4985-8a8d-68f7fd628b84/replicas/133513952307940985p, RequestStats: , SDK: Microsoft.Azure.Documents.Common/2.14.0"}, Request URI: /dbs/Pickem/colls, RequestStats: , SDK: Microsoft.Azure.Documents.Common/2.14.0, Microsoft.Azure.Documents.Common/2.14.0, Microsoft.Azure.Documents.Common/2.14.0, Microsoft.Azure.Documents.Common/2.14.0, Microsoft.Azure.Documents.Common/2.14.0
ActivityId: 353df5f2-df6d-44d1-8ac4-ddc844aef492, Microsoft.Azure.Documents.Common/2.14.0
```

Has anyone else gotten the same recommendation? What'd you do to fix it? 

For reference, the previous default index was the following.

```json
{
    "indexingMode": "consistent",
    "automatic": true,
    "includedPaths": [
        {
            "path": "/*"
        }
    ],
    "excludedPaths": [
        {
            "path": "/\"_etag\"/?"
        }
    ]
}
```

**Comment by etriebe** [comment]
@toiyabe not sure if you've gotten the same recommendation or not.

**Reply by toiyabe** [reply]
Hello Eric,

I haven't run into this yet.

Let me give this a try and get back to you.

Just so I understand you're trying to create an index on UserName on the IdentityUser collection?

Thanks,

-Eric

**Reply by etriebe** [reply]
Username and Discriminator. I don't think it let me do either of these. Thanks for taking a look!

**Comment by toiyabe** [comment]
Hello Eric,

I'm not an expert in Cosmos indexing by any means, so I'm interested in this as well.

I did some readying on Cosmos Indexes ([see link](IdentityUser)). By default Cosmos already indexes everything. I think that is what is referenced by the "default" indexing policy you have above.

Having said that, I have seen YouTube videos where people set their own.

I've seen either "*" or "?" added to the path, like this ([see example](https://learn.microsoft.com/en-us/azure/cosmos-db/index-policy#includeexclude-precedence)):

```json
{
    "indexingMode": "consistent",
    "automatic": true,
    "includedPaths": [
        {
            "path": "/Discriminator/*"
        },
        {
            "path": "/UserName/*"
        }
    ],
    "excludedPaths": [
        {
            "path": "/*"
        }
    ]
}
```

If you add this, does it work?

**Comment by etriebe** [comment]
That worked! I guess having the trailing /* made all the difference. Thank you!

**Reply by toiyabe** [reply]
Great!
I'm wondering if the /* is like a wildcard? Sort of like index the path and everything under it?
Have a nice day,
-Eric
